runtimes/web/src/framebuffer.js

```diff
@@ -29,7 +39,64 @@ export class Framebuffer {
         }
     }
 
-    drawRect (x, y, width, height) {
+    drawHLineInternal(color, x, y, len) {
+        if (y > HEIGHT || y < 0 || len === 0) {
+            return;
+        }
+        let startX, endX;
+
+        if(len < 0) {
+            startX = Math.max(0, x + len + 1);
+            endX = Math.min(WIDTH, x + 1);
+        } else {
+            startX = Math.max(0, x);
+            endX = Math.min(x + len, WIDTH);
+        }
+
+        for (let xx = startX; xx < endX; ++xx) {
+            this.drawPoint(color, xx, y);
+        }
+    }
+
+    drawVLineInternal(color, x, y, len) {
+        if (x > WIDTH || x < 0 || len === 0) {
+            return;
+        }
```

Comment:
I think drawing out of buffer should be possible. This is a fairly common situation, especially with partial visibility.
It only makes sense not to draw in such cases:
```ts
if (len === 0 || (len > 0 && x > WIDTH) || (len < 0 && x < 0)) {
  return;
}

// clamp x and len to buffer bounds
```

Reply:
It should not be possible to draw out of because `startX|Y` and `endX|Y` are bounded to `0;SIZE`. 

- https://github.com/aduros/wasm4/pull/103/files#diff-e4f17ff8a6bc469414b480445800684082740355320777c0c00297b36a91f60fR63

- https://github.com/aduros/wasm4/pull/103/files#diff-e4f17ff8a6bc469414b480445800684082740355320777c0c00297b36a91f60fR43

Is there an edge case I'm missing?

Furthermore I'd prefer not to completely skip drawing if part of the line is out of bounds.

### Example

```ts
hline(20, 20, 500);
```

Draws from (20;20) to (159; 20);



<img width="982" alt="Screenshot 2021-09-20 at 13 48 59" src="https://user-images.githubusercontent.com/16583125/133997347-c477dfa0-6d2e-4c10-a6a9-08ee015eebd2.png">

Reply:
I mean situation line this:
```
hline(-20, 20, 100);
```

Reply:
I think should be:
```
if(len === 0 || (len + x < 0)  ||  (x >= WIDTH && len > 0))
     return
```

Reply:
+1 for supporting negative X/Y. I don't think it should support negative width/height though. It's inconsistent with the rest of the drawing functions and makes things more complicated.

Reply:
>  It's inconsistent with the rest of the drawing functions and makes things more complicated.


I understand your point, my use case for negative `len` is the following:

I want draw a horizontal line from the end of screen that covers `len` pixels.

I think it is easier to write `hline(w4.WIDTH, 20, -5)` instead of writing `hline(w4.WIDTH - 4, 20,  5)`.

I'll edit this part if you deem that's better to not have a negative `len`.

Reply:
I've changed `len` to uint as suggested [here](https://github.com/aduros/wasm4/pull/103#discussion_r712208188)

@MaxGraey 

### negative -X/Y

- `hline(-20, 20, 100)`  draws a line from horizontal line between 0 and 79.

- `hline(WIDTH + 20, 20, -100)` it does not work because `len` cannot be negative.